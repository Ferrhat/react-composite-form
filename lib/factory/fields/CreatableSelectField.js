import React, {Component} from 'react';
import { Creatable } from 'react-select';
import { get, isNumber, sortBy, defaultTo } from 'lodash';

class CreatableSelectField extends Component {

    constructor(props) {
        super(props);
        this.state = {
            options: [...props.options],
        };
    }

    render() {
        const { name, label, value, valueUpdateHandler, changeHandler, disabled, options } = this.props;

        const onChange = (selection) => {
            if (get(selection, 'value', null) && !this.state.options.find(option => option.value == selection.value)) {
                this.setState({
                    options: [...this.state.options, selection],
                });
            }
            valueUpdateHandler(name, get(selection, 'value', null));
            changeHandler(selection);
        }

        return (
            <Creatable
                value={ defaultTo(value, null) }
                onChange={ onChange }
                options={ sortBy(this.state.options, 'label') }
                isLoading={!options.length}
                disabled={disabled || !options.length}
                backspaceRemoves={false}
            />
        );
    }
}

export default CreatableSelectField;
