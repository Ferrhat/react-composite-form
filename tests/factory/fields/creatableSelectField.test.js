import React from 'react';
import {shallow, mount} from 'enzyme';
import CreatableSelectField from "../../../lib/factory/fields/CreatableSelectField";


describe('CreatableSelectField', () => {
    const mockValueUpdateHandler = jest.fn();
    const mockChangeHandler = jest.fn();
    const creatableSelectField = shallow(<CreatableSelectField valueUpdateHandler={mockValueUpdateHandler} changeHandler={mockChangeHandler} options={[]} name={'testName'} />);

    it('renders properly', () => {
        expect(creatableSelectField).toMatchSnapshot();
    });

    it('stops loading if it contains options', () => {
        expect(creatableSelectField.find('CreatableSelect').prop('isLoading')).toEqual(true);
        creatableSelectField.setProps({options: [{label: 'Test', value: 1}]}, () => {
            expect(creatableSelectField.find('CreatableSelect').prop('isLoading')).toEqual(false);
        });
    });

    it('should create a new option', () => {
        const selection = {
            value: 'testValue',
        };
        creatableSelectField.find('CreatableSelect').simulate('change', selection);
        expect(mockValueUpdateHandler).toBeCalledWith('testName', 'testValue');
        expect(mockChangeHandler).toBeCalledWith(selection);
    });

    it('should select an existing option', () => {
        const selection = {
            value: 1,
        };
        creatableSelectField.find('CreatableSelect').simulate('change', selection);
        expect(mockValueUpdateHandler).toBeCalledWith('testName', 1);
        expect(mockChangeHandler).toBeCalledWith(selection);
    });

});
