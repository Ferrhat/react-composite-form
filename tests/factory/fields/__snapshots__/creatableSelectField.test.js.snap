// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatableSelectField renders properly 1`] = `
<div>
  <CreatableSelect
    backspaceRemoves={false}
    disabled={true}
    filterOptions={[Function]}
    isLoading={true}
    isOptionUnique={[Function]}
    isValidNewOption={[Function]}
    menuRenderer={[Function]}
    newOptionCreator={[Function]}
    onChange={[Function]}
    options={Array []}
    promptTextCreator={[Function]}
    shouldKeyDownEventCreateNewOption={[Function]}
    value={null}
  />
  <div
    className="validationErrorMessage"
  />
</div>
`;
